langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/Round.java

```diff
@@ -31,7 +34,39 @@
 //)
 public class Round {
 
-    public static double round(double x) {
-        return Math.rint(x);
+    public static double round(double x, long fractionDigits) {
+        if (isSpecialValue(x)) {
+            return x;
+        }
+        if (fractionDigits == 0) {
+            return Math.rint(x);
+        }
+        if (fractionDigits > Integer.MAX_VALUE) {
+            // As per IEEE, exponent of double value lies from -308 to 307.
+            // Also, the maximum decimal digits that can have in double value are 15 or 16.
+            // Therefore, if `fractionDigits` is very large, `x` will not be changed.
+            return x;
+        }
+        if (fractionDigits < Integer.MIN_VALUE) {
+            return 0;
+        }
+        // Down cast can be done safely because of above condition.
+        int fractionDigitsAsInt = (int) fractionDigits;
+        BigDecimal xInBigDecimal = BigDecimal.valueOf(x);
+        int scale = xInBigDecimal.scale();
+        if (fractionDigitsAsInt > 0) {
+            if (fractionDigitsAsInt > scale) {
+                return x;
+            }
+        } else {
+            if (-fractionDigitsAsInt > (xInBigDecimal.precision() - scale)) {
+                return 0;
+            }
+        }
+        return xInBigDecimal.setScale(fractionDigitsAsInt, RoundingMode.HALF_EVEN).doubleValue();
+    }
+
+    private static boolean isSpecialValue(double x) {
+        return Double.isInfinite(x) || Double.isNaN(x) || x == 0.0d;
```

Comment:
In this particular case, since "special value" doesn't quite convey what we're checking for and it isn't a term from the spec, I think we can just inline the check.

Reply:
Changed